controllers/vspherevm_controller.go

```diff
@@ -0,0 +1,297 @@
+			vsphereVM.Name)
+	}
+
+	// Create the VM context for this request.
+	vmContext := &context.VMContext{
+		ControllerContext: r.ControllerContext,
+		VSphereVM:         vsphereVM,
+		Session:           authSession,
+		Logger:            r.Logger.WithName(req.Namespace).WithName(req.Name),
+		PatchHelper:       patchHelper,
+	}
+
+	// Print the task-ref upon entry and upon exit.
+	vmContext.Logger.V(4).Info(
+		"VSphereVM.Status.TaskRef OnEntry",
+		"task-ref", vmContext.VSphereVM.Status.TaskRef)
+	defer func() {
+		vmContext.Logger.V(4).Info(
+			"VSphereVM.Status.TaskRef OnExit",
+			"task-ref", vmContext.VSphereVM.Status.TaskRef)
+	}()
+
+	// Always issue a patch when exiting this function so changes to the
+	// resource are patched back to the API server.
+	defer func() {
+		// Patch the VSphereVM resource.
+		if err := vmContext.Patch(); err != nil {
+			if reterr == nil {
+				reterr = err
+			}
+			vmContext.Logger.Error(err, "patch failed", "vm", vmContext.String())
+		}
+
+		// localObj is a deep copy of the VSphereVM resource that was
+		// fetched at the top of this Reconcile function.
+		localObj := vmContext.VSphereVM.DeepCopy()
+
+		// Fetch the up-to-date VSphereVM resource into remoteObj until the
+		// fetched resource has a a different ResourceVersion than the local
+		// object.
+		//
+		// FYI - resource versions are opaque, numeric strings and should not
+		// be compared with < or >, only for equality -
+		// https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions.
+		//
+		// Since CAPV is currently deployed with a single replica, and this
+		// controller has a max concurrency of one, the only agent updating the
+		// VSphereVM resource should be this controller.
+		//
+		// So if the remote resource's ResourceVersion is different than the
+		// ResourceVersion of the resource fetched at the beginning of this
+		// reconcile request, then that means the remote resource should be
+		// newer than the local resource.
+		wait.PollImmediateInfinite(time.Second*1, func() (bool, error) {
+			// remoteObj refererences the same VSphereVM resource as it exists
+			// on the API server post the patch operation above. In a perfect world,
+			// the Status for localObj and remoteObj should be the same.
+			remoteObj := &infrav1.VSphereVM{}
+			if err := vmContext.Client.Get(vmContext, req.NamespacedName, remoteObj); err != nil {
+				if apierrors.IsNotFound(err) {
+					// It's possible that the remote resource cannot be found
+					// because it has been removed. Do not error, just exit.
+					return true, nil
+				}
+
+				// There was an issue getting the remote resource. Sleep for a
+				// second and try again.
+				vmContext.Logger.Error(err, "failed to get VSphereVM while exiting reconcile")
+				return false, nil
+			}
+
+			// If the remote resource version is not the same as the local
+			// resource version, then it means we were able to get a resource
+			// newer than the one we already had.
```

Comment:
To match the comment, shouldn't the condition below be changed to:

localObj.ResourceVersion < remoteObj.ResourceVersion

Reply:
Hi @tedyu,

One would think, but see the above comment:

> ```golang
> 		// FYI - resource versions are opaque, numeric strings and should not
>  		// be compared with < or >, only for equality -
>  		// https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions.
>  		//
>  		// Since CAPV is currently deployed with a single replica, and this
>  		// controller has a max concurrency of one, the only agent updating the
>  		// VSphereVM resource should be this controller.
>  		//
>  		// So if the remote resource's ResourceVersion is different than the
>  		// ResourceVersion of the resource fetched at the beginning of this
>  		// reconcile request, then that means the remote resource should be
>  		// newer than the local resource.
> ```